### PR TITLE
Provide unit tests for flags to configure secure connection client certificate

### PR DESCRIPTION
**Pull request by gboyvalenkov-bosch**
[#14] Provide unit tests for flags to configure secure connection client certificate

- unit tests implemented

Signed-off-by: Georgi Boyvalenkov <Georgi.Boyvalenkov@bosch.io>

### CODE REVIEW

File: internal/feature_test.go [hunk 1]
```diff
@@ -96,38 +100,48 @@ func TestScriptBasedInit(t *testing.T) {
...
+		clientConnected: false, storageLocation: dir, featureID: getDefaultFlagValue(t, flagFeatureID)})
 	if err == nil {
-		t.Fatal("Ditto Client shall not be connected!")
+		t.Fatal("ditto Client shall not be connected!")
 	}
 
 }
 
-// TestSBSUInit tests ScriptBasedSoftwareUpdatable core functionality: init, install and download.
+// TestScriptBasedCore tests ScriptBasedSoftwareUpdatable core functionality: init, install and download.
```
Comment (e-grigorov) [inline internal/feature_test.go:110]: TestScriptBasedCore - actually, it's software update core.

File: internal/storage/download_test.go [hunk 2]
```diff
@@ -181,119 +222,67 @@ func TestDownloadToFileError(t *testing.T) {
...
-	begin, end := int64(0), int64(w.size-1)
-	r = r[6:] // Strip the "bytes=", left over is now "begin-end"
-	pos := strings.Index(r, "-")
-	if pos < 0 {
-		http.Error(writer, "invalid values for header 'Range'", http.StatusBadRequest)
-		return
+	// 1. Try download with expired certificate
+	art.Link = "https://localhost:43234/test.txt"
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:264]: The message can be done more specific for this case where the client is not using the expired certificate.

File: internal/storage/download_test.go [hunk 3]
```diff
@@ -181,119 +222,67 @@ func TestDownloadToFileError(t *testing.T) {
...
-
-	// Parse begin part of the range
-	sb := r[:pos]
-	if len(sb) > 0 {
-		begin, err = strconv.ParseInt(sb, 10, 64)
-		if err != nil || begin > w.size {
-			http.Error(writer, "invalid values for header 'Range' begin", http.StatusBadRequest)
-		}
+	if err := downloadArtifact(name, art, nil, expiredCert, make(chan struct{})); err == nil {
+		t.Fatalf("validated with expired certificate: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:267]: The message can be done more specific for this case where the client and server are using the expired certificate.

File: internal/storage/download_test.go [hunk 4]
```diff
@@ -181,119 +222,67 @@ func TestDownloadToFileError(t *testing.T) {
...
-	se := r[pos+1:]
-	if len(se) > 0 {
-		end, err = strconv.ParseInt(se, 10, 64)
-		if err != nil || end > w.size || begin >= end {
-			http.Error(writer, "invalid values for header 'Range' end", http.StatusBadRequest)
-		}
+	// 2. Try download with untrusted certificate
+	art.Link = "https://localhost:43235/test.txt"
+	if err := downloadArtifact(name, art, nil, "", make(chan struct{})); err == nil {
+		t.Fatalf("validated with untrusted certificate: %v", err)
```
Comment (e-grigorov) [inline internal/storage/download_test.go:273]: How "ext" relates with untrusted?

File: internal/storage/download_test.go [hunk 5]
```diff
@@ -181,119 +222,67 @@ func TestDownloadToFileError(t *testing.T) {
...
-	write(writer, end-begin+1)
-}
-
-// handlerSimple handles incoming HTTP requests without range header support.
-func (w *web) handlerSimple(writer http.ResponseWriter, request *http.Request) {
-	writer.Header().Set("Content-Type", "text/plain")
-	writer.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="%s"`, w.name))
-	writer.Header().Set("Content-Length", strconv.Itoa(int(w.size)))
-	write(writer, w.size)
+	// 3. Try download with unknown certificate
```
Comment (e-grigorov) [inline internal/storage/download_test.go:276]: The comment is not correct, it tries to download with untrusted certificate.

File: internal/storage/web_test_util.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/eclipse-kanto/software-update/hawkbit"
+)
+
+// Web encapsulates a test HTTP(S) server and testing data
+type Web struct {
```
Comment (e-grigorov) [inline internal/storage/web_test_util.go:32]: Web (file name, struct etc.) doesn't sound correct. Isn't it a test HTTP server?

**Review by e-grigorov — CHANGES_REQUESTED** [review]

File: internal/feature_test.go (hunk 1 above)
Comment (gboyvalenkov-bosch) [reply internal/feature_test.go:110]: The naming should be fixed in another PR.

File: internal/storage/http_test_util.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/eclipse-kanto/software-update/hawkbit"
+)
+
+// Web encapsulates a test HTTP(S) server and testing data
+type Web struct {
```
Comment (e-grigorov) [inline internal/storage/http_test_util.go:32]: It's still web, I would suggest to align with HTTP server or something similar.

File: internal/storage/http_test_util.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+		testOnceSimple.Do(func() { http.HandleFunc(fmt.Sprintf("/%s", name), w.handlerSimple) })
+	} else {
+		testOnceRange.Do(func() { http.HandleFunc(fmt.Sprintf("/%s", name), w.handlerRange) })
+	}
+
+	// Start HTTP server in separate goroute.
+	if secure {
+		go func() {
+			if err := w.srv.ListenAndServeTLS(cert, key); err != nil && err != http.ErrServerClosed {
+				t.Errorf("Failed to start secure web server: %v", err)
```
Comment (e-grigorov) [inline internal/storage/http_test_util.go:78]: Align "web" with the new name.

File: internal/storage/http_test_util.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+			}
+		}()
+	} else {
+		go func() {
+			if err := w.srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+				t.Errorf("Failed to start web server: %v", err)
```
Comment (e-grigorov) [inline internal/storage/http_test_util.go:84]: Align "web" with the new name.

**Review by e-grigorov — CHANGES_REQUESTED** [review]

**Review by e-grigorov — APPROVED** [review]

**Pull request merged**